File: card-check-read-as-a-failed-payment/node/card-check-auditor.js
/**
 * Fix subscriptions dunned for a zero amount Stripe card check, not a real failed renewal.
 *
 * Stripe sometimes verifies a saved card with a $0 PaymentIntent, for example after a
 * card updater event or a trial signup. If that check does not come back clean, some
 * failure handling treats it exactly like a declined renewal charge. This walks recently
 * dunned subscriptions, reads the PaymentIntent behind the last order, and reactivates
 * any subscription whose "failure" was really a zero amount check.
 *
 * Guide: https://www.allanninal.dev/woocommerce/card-check-read-as-a-failed-payment/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 3);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const FAILED_STATUSES = new Set(["on-hold", "pending-cancel"]);

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

/**
 * Pure decision: should a dunned subscription be restored because the "failed"
 * intent behind it was really a zero amount card check?
 *
 * subscription: object with at least a `status` property.
 * intent: a Stripe PaymentIntent-like object (or null/undefined if there was
 *   nothing to check), read for `amount` (minor units) and `status`.
 * Returns a [action, reason] tuple. action is one of "restore" or "skip".
 */
export function decide(subscription, intent) {
  if (!FAILED_STATUSES.has(subscription.status)) return ["skip", "subscription is not in a dunned state"];
  if (!intent) return ["skip", "no Stripe intent to check"];
  if ((intent.amount || 0) === 0) return ["restore", "the failed intent was a zero amount card check"];
  if (intent.status === "succeeded") return ["skip", "the intent actually succeeded, nothing to fix"];
  return ["skip", "a real charge was attempted and declined"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getOrder(orderId) {
  if (!orderId) return null;
  return woo(`/orders/${orderId}`);
}

async function getIntent(intentId) {
  if (!intentId) return null;
  try {
    return await stripe.paymentIntents.retrieve(intentId);
  } catch {
    return null;
  }
}

async function* heldSubscriptions() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/subscriptions?status=on-hold,pending-cancel&modified_after=${after}&per_page=50&page=${page}`);
    if (!batch || !batch.length) return;
    for (const sub of batch) yield sub;
    page++;
  }
}

async function restore(subscriptionId, intent) {
  await woo(`/subscriptions/${subscriptionId}`, {
    method: "PUT",
    body: JSON.stringify({ status: "active" }),
  });
  await woo(`/subscriptions/${subscriptionId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Reactivated by the card check auditor. PaymentIntent ${intent.id} ` +
            `had amount 0, a card check, not a failed renewal. No dunning is owed here.`,
    }),
  });
}

export async function run() {
  let restored = 0;
  for await (const subscription of heldSubscriptions()) {
    const orderId = subscription.last_order_id || subscription.last_order;
    const order = await getOrder(orderId);
    const intent = order ? await getIntent(intentIdOf(order)) : null;
    const [action, reason] = decide(subscription, intent);
    if (action !== "restore") continue;
    console.log(`Subscription ${subscription.id}: ${reason}. ${DRY_RUN ? "would restore" : "restoring"}`);
    if (!DRY_RUN) await restore(subscription.id, intent);
    restored++;
  }
  console.log(`Done. ${restored} subscription(s) ${DRY_RUN ? "to restore" : "restored"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: card-check-read-as-a-failed-payment/python/card_check_auditor.py
"""Fix subscriptions dunned for a zero amount Stripe card check, not a real failed renewal.

Stripe sometimes verifies a saved card with a $0 PaymentIntent, for example after a card
updater event or a trial signup. If that check does not come back clean, some failure
handling treats it exactly like a declined renewal charge. This walks recently dunned
subscriptions, reads the PaymentIntent behind the last order, and reactivates any
subscription whose "failure" was really a zero amount check. Read only by default is not
possible here (the point is to restore access), so it defaults to a dry run instead.

Guide: https://www.allanninal.dev/woocommerce/card-check-read-as-a-failed-payment/
"""
import os
import datetime
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("card_check_auditor")

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "3"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

FAILED_STATUSES = {"on-hold", "pending-cancel"}


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def decide(subscription, intent):
    """Pure decision: should a dunned subscription be restored because the
    "failed" intent behind it was really a zero amount card check?

    subscription: dict with at least a "status" key.
    intent: a Stripe PaymentIntent-like dict (or None if there was nothing to check),
        read for "amount" (minor units) and "status".
    Returns a (action, reason) tuple. action is one of "restore" or "skip".
    """
    if subscription["status"] not in FAILED_STATUSES:
        return ("skip", "subscription is not in a dunned state")
    if intent is None:
        return ("skip", "no Stripe intent to check")
    if intent.get("amount", 0) == 0:
        return ("restore", "the failed intent was a zero amount card check")
    if intent.get("status") == "succeeded":
        return ("skip", "the intent actually succeeded, nothing to fix")
    return ("skip", "a real charge was attempted and declined")


def get_order(order_id):
    if not order_id:
        return None
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def get_intent(intent_id):
    if not intent_id:
        return None
    try:
        return stripe.PaymentIntent.retrieve(intent_id)
    except stripe.error.InvalidRequestError:
        return None


def held_subscriptions():
    after = (datetime.date.today() - datetime.timedelta(days=LOOKBACK_DAYS)).isoformat() + "T00:00:00"
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/subscriptions",
            params={"status": "on-hold,pending-cancel", "modified_after": after, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        yield from batch
        page += 1


def restore(subscription_id, intent):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}",
        json={"status": "active"}, auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}/notes",
        json={"note": f"Reactivated by the card check auditor. PaymentIntent {intent['id']} "
                      f"had amount 0, a card check, not a failed renewal. No dunning is owed here."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    restored = 0
    for subscription in held_subscriptions():
        last_order_id = subscription.get("last_order_id") or subscription.get("last_order")
        order = get_order(last_order_id)
        intent = get_intent(intent_id_of(order)) if order else None
        action, reason = decide(subscription, intent)
        if action != "restore":
            continue
        log.info("Subscription %s: %s. %s", subscription["id"], reason, "would restore" if DRY_RUN else "restoring")
        if not DRY_RUN:
            restore(subscription["id"], intent)
        restored += 1
    log.info("Done. %d subscription(s) %s.", restored, "to restore" if DRY_RUN else "restored")


if __name__ == "__main__":
    run()
